crates/pipeline-manager/src/compiler/rust_compiler.rs

```diff
@@ -0,0 +1,1055 @@
+        ("platform_version", platform_version.as_bytes()),
+        ("profile", profile.to_string().as_bytes()),
+        (
+            "config.compiler_working_directory",
+            config.compiler_working_directory.to_string().as_bytes(),
+        ),
+        (
+            "config.sql_compiler_home",
+            config.sql_compiler_home.to_string().as_bytes(),
+        ),
+        (
+            "config.dbsp_override_path",
+            config.dbsp_override_path.to_string().as_bytes(),
+        ),
+        ("program_info.main_rust", main_rust.as_bytes()),
+        ("program_info.udf_stubs", udf_stubs.as_bytes()),
+        ("udf_rust", udf_rust.as_bytes()),
+        ("udf_toml", udf_toml.as_bytes()),
+    ] {
+        let checksum = sha256(data);
+        hasher.update(&checksum);
+        trace!(
+            "Rust compilation: checksum of {name}: {}",
+            hex::encode(checksum)
+        );
+    }
+    Ok(hex::encode(hasher.finish()))
+}
+
+/// Rust compilation possible error outcomes.
+pub enum RustCompilationError {
+    /// In the meanwhile the pipeline was already updated, as such the
+    /// Rust compilation is outdated and no longer useful.
+    Outdated,
+    /// The Rust compilation process was terminated by a signal.
+    /// This can happen for instance when the compiler server is terminated by signal,
+    /// and processes started by it are first terminated before itself. The signal is likely not
+    /// related to the program itself inherently being unable to compile, nor the compiler
+    /// server reaching an inconsistent state. As such, retrying is the desired
+    /// behavior rather than declaring failure to compile the specific program.
+    TerminatedBySignal,
+    /// Rust compilation was unable to be performed.
+    /// This is either due to:
+    /// - A check failed when setting up the working directory in which
+    ///   the Rust compiler will be called
+    /// - The Rust compiler call failed and returned an error
+    ///   (e.g., a syntax error in the SQL-generated Rust or
+    ///   the user-provided Rust implementation of UDFs).
+    RustError(String),
+    /// General system problem occurred (e.g., I/O error)
+    SystemError(String),
+}
+
+/// Common errors are system errors during Rust compilation.
+impl From<CommonError> for RustCompilationError {
+    fn from(value: CommonError) -> Self {
+        RustCompilationError::SystemError(value.to_string())
+    }
+}
+
+/// Performs the Rust compilation:
+/// - Calculate source checksum across fields for cache lookup
+/// - Perform cache lookup using the source checksum:
+///   - If a binary already exists for that source checksum, return that
+///   - Otherwise, set up the working directory for the Rust compiler
+///     and call it in order to generate the binary
+///
+/// Returns the program binary URL, source checksum, integrity checksum,
+/// duration and whether it was cached.
+#[allow(clippy::too_many_arguments)]
```

Comment:
maybe we need a UDF struct

Reply:
They're currently individual fields stored, and at least currently there will be no more UDF fields that will be added, hence I would keep them separate. In general, we could consider a "program" struct at some point, depending on how complex of an object that will evolve to become as a whole.